### DISCUSSION TITLE
How to capature still images?

### DISCUSSION THREAD

**Discussion by nandgator** [whitphx/streamlit-webrtc#236] [Q&A]
Here's part of my code:

```python
...
capture_img = None
class OpenCVVideoProcessor(VideoProcessorBase):
   # Inherited class
    def recv(self, frame: av.VideoFrame) -> av.VideoFrame:
        # Overridden function
        global capture_img
        capture_img = frame.to_ndarray(format='bgr24')
        return av.VideoFrame.from_ndarray(capture_img, format="bgr24")


col1, col2 = st.beta_columns(spec=2)
with col1:
    # 1. Input
    st.header(body='Input')
    webrtc_streamer(
        key='Input', video_processor_factory=OpenCVVideoProcessor)

...
```
I hoped that the `frame` converted to `ndarray` would be stored in `capture_img` since I've already started the camera.

But even though `capture_img = frame.to_ndarray(format='bgr24')` gets executed, `capture_img` stays `None`. Because when a button is pressed `capture_img` is directly assigned to some and value it raises something like `AttributeError: 'NoneType' object has no attribute 'shape'`.

Why is it so? Is there another way to capture a frame/image?

> [Full code](https://github.com/joe733/vudoku/blob/main/src/main.py)
> [Streamlit deployment](https://share.streamlit.io/joe733/vudoku/main/src/main.py) is broken due to dependency issue, everything works on local setup.

**Comment by whitphx** [comment]
Please see https://discuss.streamlit.io/t/new-component-streamlit-webrtc-a-new-way-to-deal-with-real-time-media-streams/8669/23?u=whitphx.
This post may be an answer while `transform()` was used instead of `recv()` because the post was originally written when the version of `streamlit-webrtc` was less than v0.20.

The point is that you cannot use `global` for that purpose and should pass values between the global scope and `recv()` through instance attributes of a VideoProcessor class in a thread-safe way.

This is because
* `recv()` is running in another thread
* The `capture_img` variable you defined in the global scope *changes* over executions though the reference of `global capture_img`  does not.
  * https://blog.streamlit.io/how-to-build-the-streamlit-webrtc-component/#the-basics-of-streamlit-s-execution-model may be nice to read to know about Streamlit's execution model where the script is run many times.
